src/Components/WebAssembly/WebAssembly/src/Services/Logging/WebAssemblyLoggerFactory.cs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Services
+{
+    /// <summary>
+    /// Produces instances of <see cref="ILogger"/> classes based on the given providers.
+    /// </summary>
+    public class WebAssemblyLoggerFactory : ILoggerFactory
+    {
+        private readonly Dictionary<string, Logger> _loggers = new Dictionary<string, Logger>();
+        private readonly List<ProviderRegistration> _providerRegistrations = new List<ProviderRegistration>();
+        private readonly object _sync = new object();
+        private volatile bool _disposed;
+
+        public WebAssemblyLoggerFactory() : this(Enumerable.Empty<ILoggerProvider>()) { }
+
+        /// <summary>
+        /// Creates a new <see cref="WebAssemblyLoggerFactory"/> instance.
+        /// </summary>
+        /// <param name="providers">The providers to use in producing <see cref="ILogger"/> instances.</param>
+        public WebAssemblyLoggerFactory(IEnumerable<ILoggerProvider> providers)
+        {
+            foreach (var provider in providers)
+            {
+                AddProviderRegistration(provider, dispose: false);
+            }
+        }
+
+        /// <summary>
+        /// Creates an <see cref="ILogger"/> with the given <paramref name="categoryName"/>.
+        /// </summary>
+        /// <param name="categoryName">The category name for messages produced by the logger.</param>
+        /// <returns>The <see cref="ILogger"/> that was created.</returns>
+        public ILogger CreateLogger(string categoryName)
+        {
+            if (CheckDisposed())
+            {
+                throw new ObjectDisposedException(nameof(WebAssemblyLoggerFactory));
+            }
+
+            lock (_sync)
+            {
+                if (!_loggers.TryGetValue(categoryName, out var logger))
+                {
+                    logger = new Logger
+                    {
+                        Loggers = CreateLoggers(categoryName),
+                    };
+
+                    _loggers[categoryName] = logger;
+                }
+
+                return logger;
+            }
+        }
+
+        /// <summary>
+        /// Adds the given provider to those used in creating <see cref="ILogger"/> instances.
+        /// </summary>
+        /// <param name="provider">The <see cref="ILoggerProvider"/> to add.</param>
+        public void AddProvider(ILoggerProvider provider)
+        {
+            if (CheckDisposed())
+            {
+                throw new ObjectDisposedException(nameof(WebAssemblyLoggerFactory));
+            }
+
+            lock (_sync)
```

Comment:
nit: We don't need the locks/concurrent stuff since blazor is single threaded.

Reply:
That's true today but there is a threaded version of Mono WebAssembly which we may enable use of in the near future. At that point we'll need to go through the code and make sure we're not relying on single-threadedness anywhere. So in cases where we know we'd need it, probably best to have the locks in place already :)